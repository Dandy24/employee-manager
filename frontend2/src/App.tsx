import React from 'react';
import 'antd/dist/antd.css';
import { Layout } from 'antd';
import { Route, Switch } from 'react-router-dom';

import { CompanyListPage } from './pages/CompanyList';
import { NewCompanyPage } from './pages/NewCompany';
import { MainNavigation } from './components/layout/MainNavigation';
import { NewEmployeePage } from './pages/NewEmployee';
import { EmployeeListPage } from './pages/EmployeeList';
import { Content } from 'antd/lib/layout/layout';
import { useRootStore } from './stores/root-store-provider';
import { observer } from 'mobx-react-lite';
import { ShiftManagerPage } from './pages/ShiftManager';

const App: React.FC = observer(() => {
    const rootStore = useRootStore();

    return (
        <>
            <Layout>
                <MainNavigation />
                <Content style={{ padding: '35px 50px' }}>
                    <Switch>
                        <Route path="/" exact>
                            <CompanyListPage rootStore={rootStore}></CompanyListPage>
                        </Route>

                        <Route path="/new-company">
                            <NewCompanyPage rootStore={rootStore}></NewCompanyPage>
                        </Route>

                        <Route path="/new-employee">
                            <NewEmployeePage rootStore={rootStore}></NewEmployeePage>
                        </Route>

                        <Route path="/employee-list">
                            <EmployeeListPage rootStore={rootStore}></EmployeeListPage>
                        </Route>

                        <Route path="/shift-manager">
                            <ShiftManagerPage rootStore={rootStore}></ShiftManagerPage>
                        </Route>

                        {/*<Route path="/monthly-output/:id" exact>*/}
                        {/*    <MonthlyReview></MonthlyReview>*/}
                        {/*</Route>*/}
                    </Switch>
                </Content>
            </Layout>
        </>
    );
});

export default App;
