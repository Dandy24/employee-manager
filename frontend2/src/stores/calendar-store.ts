import { RootStore } from './root-store';
import { action, computed, makeObservable, observable } from 'mobx';
import moment from 'moment';

export class CalendarStore {
    selectedDate: moment.Moment;

    activeCompanyId: number;

    isEditOpen = false;
    isShiftSelectOpen = false;

    private rootStore: RootStore;

    constructor(rootStore: RootStore) {
        this.rootStore = rootStore;

        makeObservable(this, {
            isShiftSelectOpen: observable,
            isEditOpen: observable,
            setShiftSelectOpen: action,
            setShiftEditOpen: action,

            activeCompanyId: observable,
            setActiveCompanyId: action,

            selectedDate: observable,
            setSelectedDate: action,

            formattedDate: computed,
            stringDate: computed,
        });
    }

    setActiveCompanyId(companyId: number): void {
        this.activeCompanyId = companyId;
    }

    setShiftSelectOpen(open: boolean): void {
        this.isShiftSelectOpen = open;
    }

    setShiftEditOpen(open: boolean): void {
        this.isEditOpen = open;
    }

    setSelectedDate(date: moment.Moment): void {
        if (date) {
            this.selectedDate = date;
        }
    }

    get formattedDate(): string {
        const day = this.selectedDate?.format('dddd');
        const date = this.selectedDate?.format('LL');
        return `${day} ${date}`;
    }

    get stringDate(): string {
        return moment(this.selectedDate).format('YYYY-MM-DD');
    }
}
