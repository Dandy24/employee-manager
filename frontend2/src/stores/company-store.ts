import { RootStore } from './root-store';
import { action, makeObservable, observable, runInAction } from 'mobx';
import { createCompany, deleteCompany, getCompanyList, updateCompany } from '../api/apiCalls';
import { message } from 'antd';
import { CompanyEntity } from '../models/entities/company-entity';
import { CompanyDto } from '../models/dtos/company-dto';
import { SearchableCompanyEntity } from '../models/entities/searchable-company-entity';

export class CompanyStore {
    companies: CompanyEntity[] = [];
    company: CompanyEntity;
    loadingCompanies = false;
    isEditOpen = false;
    rootStore: RootStore;

    constructor(rootStore: RootStore) {
        this.rootStore = rootStore;

        makeObservable(this, {
            company: observable,
            companies: observable,
            loadingCompanies: observable,
            isEditOpen: observable,

            fetchAllCompanies: action,
            closeModal: action,
            openToEdit: action,
            addCompany: action,
            editCompany: action,
            updateCompany: action,
            deleteCompany: action,
        });
    }

    async fetchAllCompanies(filter: SearchableCompanyEntity = null): Promise<void> {
        runInAction(() => {
            this.loadingCompanies = true;
            this.companies = [];
        });

        const companies = await getCompanyList();

        let filteredCompanies;

        if (filter?.company) {
            filteredCompanies = companies.filter(
                (comp) => comp.name === filter.company.name || comp.address === filter.company.address,
            );
        }

        runInAction(() => {
            this.loadingCompanies = false;
            this.companies = filteredCompanies ? filteredCompanies : companies;
        });
    }

    closeModal(): void {
        this.isEditOpen = false;
    }

    openToEdit(company: CompanyEntity): void {
        const comp = this.companies.find((c) => c.id === company.id);
        this.isEditOpen = true;
        if (comp) {
            runInAction(() => {
                this.company = comp;
            });
        } else {
            throw new Error(`Spolecnost s ID ${company.id} nenalezena!`);
        }
    }

    async editCompany(company: CompanyDto): Promise<void> {
        if (this.company.id) {
            const updatedCompany = await updateCompany(company, this.company.id)
                .then(() => {
                    message.warning('Údaje o společnosti byly upraveny');
                })
                .catch((error) => {
                    message.error('Údaje o společnosti se nepodařilo upravit.');
                    console.log(error);
                }); //TODO dat to ID pryc nejak
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            if (updatedCompany) {
                this.updateCompany(updatedCompany);
            }
            await this.fetchAllCompanies();
        }

        runInAction(() => {
            this.closeModal();
        });
    }

    async addCompany(company: CompanyDto): Promise<void> {
        await createCompany(company)
            .then(() => {
                message.success('Spolecnost byla uspesne vytvorena');
            })
            .catch((error) => {
                message.error('Spolecnost se nepodarilo vytvorit');
                console.log(error);
            });
        await this.fetchAllCompanies();
        runInAction(() => {
            this.closeModal();
        });
    }

    async deleteCompany(company: CompanyEntity): Promise<void> {
        await deleteCompany(company.id).catch(() => {
            message.error('Firmu se nepodařilo smazat.');
        });
        runInAction(() => {
            this.loadingCompanies = true;
            message.success('Firma byla úspěšně smazána.');
        });
        await this.fetchAllCompanies();
        runInAction(() => {
            this.loadingCompanies = false;
        });
    }

    updateCompany(updatedCompany: CompanyEntity): void {
        runInAction(() => {
            const updatedCompanyIndex = this.companies.findIndex((company) => company.id === updatedCompany.id);
            if (updatedCompanyIndex !== -1) {
                this.companies[updatedCompanyIndex] = updatedCompany;
            }
        });
    }
}
