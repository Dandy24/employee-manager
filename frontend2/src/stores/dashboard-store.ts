import { RootStore } from './root-store';
import { action, computed, makeObservable, observable, runInAction } from 'mobx';
import { EmployeeMonthlyOutputEntity } from '../models/entities/employee-monthly-output-entity';
import {
    getEmployeeMonthlyOutput,
    getEmployeeTopMonthsList,
    getMonthlyHoursByCompany,
    getOverallMonthlyOutput,
    getTopEmployeeOutputList,
} from '../api/apiCalls';
import { message } from 'antd';
import { OverallMonthlyOutputEntity } from '../models/entities/overall-monthly-output-entity';
import {
    ExtendedHoursTypeGraphDataInterface,
    GraphDataInterface,
    HoursTypeGraphDataInterface,
} from '../models/interfaces/graph-data-interface';
import moment from 'moment';

export class DashboardStore {
    employeeOutput: EmployeeMonthlyOutputEntity[];
    overallOutput: OverallMonthlyOutputEntity[];
    topEmployees: EmployeeMonthlyOutputEntity[];
    topEmployeeMonths: EmployeeMonthlyOutputEntity[];
    employeeMode = false;
    companyHours: GraphDataInterface[];
    private rootStore: RootStore;

    constructor(rootStore: RootStore) {
        this.rootStore = rootStore;

        makeObservable(this, {
            employeeOutput: observable,
            overallOutput: observable,
            topEmployees: observable,
            topEmployeeMonths: observable,
            employeeMode: observable,
            companyHours: observable,
            loadEmployeeOutput: action,
            loadOverallOutput: action,
            loadHoursByCompany: action,
            loadTopEmployeesOutputList: action,
            loadEmployeesTopMonthsList: action,
            switchMode: action,

            workingDaysGraphData: computed,
            effectivityGraphData: computed,
            hoursDistributionGraphData: computed,
            housingCapacity: computed,
            overallEffectivity: computed,
            topEmployeeOutputsData: computed,
            employeeTopMonthsOutputsData: computed,
        });
    }

    switchMode(): void {
        this.employeeMode = !this.employeeMode;
    }

    async loadEmployeeOutput(id: number): Promise<void> {
        runInAction(() => {
            // this.loadingCompanies = true;
            this.employeeOutput = [];
        });

        let output = [];

        try {
            output = await getEmployeeMonthlyOutput(id);
        } catch (e) {
            message.error('Failed to load employee output from database');
        } finally {
            // this.loadingCompanies = false;
            runInAction(() => {
                this.employeeOutput = output;
            });
        }
    }

    async loadOverallOutput(): Promise<void> {
        runInAction(() => {
            // this.loadingCompanies = true;
            this.overallOutput = [];
        });

        let output = [];

        try {
            output = await getOverallMonthlyOutput();
        } catch (e) {
            message.error('Failed to load overall output from database');
        } finally {
            // this.loadingCompanies = false;
            runInAction(() => {
                this.overallOutput = output;
            });
        }
    }

    async loadHoursByCompany(): Promise<void> {
        runInAction(() => {
            // this.loadingCompanies = true;
            this.companyHours = [];
        });

        let output = [];
        const formattedOutput = [];

        try {
            output = await getMonthlyHoursByCompany(this.overallOutput[0].start_date, this.overallOutput[0].end_date);

            output.forEach((out) => {
                const comp = { name: out[0], value: out[1] };
                formattedOutput.push(comp);
            });
        } catch (e) {
            message.error('Failed to load overall output from database');
        } finally {
            // this.loadingCompanies = false;
            runInAction(() => {
                this.companyHours = formattedOutput;
            });
        }
    }

    async loadTopEmployeesOutputList(): Promise<void> {
        runInAction(() => {
            // this.loadingCompanies = true;
            this.topEmployees = [];
        });

        let output = [];

        try {
            output = await getTopEmployeeOutputList();
        } catch (e) {
            message.error('Failed to load top employees output list from database');
        } finally {
            // this.loadingCompanies = false;
            runInAction(() => {
                this.topEmployees = output;
            });
        }
    }

    async loadEmployeesTopMonthsList(id: number): Promise<void> {
        runInAction(() => {
            // this.loadingCompanies = true;
            this.topEmployeeMonths = [];
        });

        let output = [];

        try {
            output = await getEmployeeTopMonthsList(id);
        } catch (e) {
            message.error('Failed to load top employees output list from database');
        } finally {
            // this.loadingCompanies = false;
            runInAction(() => {
                this.topEmployeeMonths = output;
            });
        }
    }

    get workingDaysGraphData(): HoursTypeGraphDataInterface[] {
        return this.employeeOutput
            ?.map((output) => ({
                name: output.start_date,
                work: output.working_hours,
                vac: output.vacation_hours,
            }))
            .reverse();
    }

    get effectivityGraphData(): GraphDataInterface[] {
        return this.employeeOutput
            .map((output) => ({
                name: output.start_date,
                value: output.effectivity,
            }))
            .reverse();
    }

    get hoursDistributionGraphData(): GraphDataInterface[] {
        return [
            { name: 'Hours worked', value: this.employeeOutput[0]?.working_hours },
            { name: 'Hours vacation', value: this.employeeOutput[0]?.vacation_hours },
            { name: 'Hours sick', value: this.employeeOutput[0]?.sick_hours },
            { name: 'Hours overtime', value: this.employeeOutput[0]?.overtime_hours },
        ];
    }

    get overallEffectivity(): GraphDataInterface[] {
        return this.overallOutput
            .map((output) => ({
                name: output.start_date,
                value: Math.ceil(output.effectivity),
            }))
            .reverse();
    }

    get housingCapacity(): number {
        return Math.ceil((this.overallOutput[0]?.housing_capacity / 15) * 100);
    }

    get overallWorkingDaysGraphData(): HoursTypeGraphDataInterface[] {
        return this.overallOutput.map((output) => ({
            name: output.start_date,
            work: output.working_hours,
            vac: output.vacation_hours,
        }));
    }

    // FIXME first_name + last_name
    get topEmployeeOutputsData(): ExtendedHoursTypeGraphDataInterface[] {
        return this.topEmployees?.map((output) => ({
            id: output.employee,
            name: this.rootStore.employeeStore.employees.find((emp) => emp.id === output.employee)?.last_name,
            work: output.working_hours,
            vac: output.vacation_hours,
            sick: output.sick_hours,
            overtime: output.overtime_hours,
        }));
    }

    get employeeTopMonthsOutputsData(): ExtendedHoursTypeGraphDataInterface[] {
        return this.topEmployeeMonths?.map((output) => ({
            name: moment(output.start_date, 'YYYY-MM-DD').format('MMMM YY'),
            work: output.working_hours,
            vac: output.vacation_hours,
            sick: output.sick_hours,
            overtime: output.overtime_hours,
        }));
    }
}
