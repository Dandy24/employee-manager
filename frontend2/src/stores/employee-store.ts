import { RootStore } from './root-store';
import { action, makeObservable, observable, runInAction } from 'mobx';
import {
    createEmployee,
    deleteEmployee,
    getEmployeeList,
    getEmployeeListForCompany,
    updateEmployee,
} from '../api/apiCalls';
import { message } from 'antd';
import { EmployeeDto } from '../models/dtos/employee-dto';
import { EmployeeEntity } from '../models/entities/employee-entity';

export class EmployeeStore {
    employees: EmployeeEntity[] = [];
    employee: any = {};
    loadingEmployees = false;
    isEditOpen = false;
    private rootStore: RootStore;

    constructor(rootStore: RootStore) {
        this.rootStore = rootStore;

        makeObservable(this, {
            employees: observable,
            employee: observable,
            loadingEmployees: observable,
            isEditOpen: observable,

            fetchAllEmployees: action,
            closeModal: action,
            openToEdit: action,
            addEmployee: action,
            editEmployee: action,
            deleteEmployee: action,
        });
    }

    async fetchAllEmployees(companyId: number = null, shiftId: number = null): Promise<void> {
        runInAction(() => {
            this.loadingEmployees = true;
            this.employees = [];
        });
        if (companyId) {
            await getEmployeeListForCompany(companyId).then((data) =>
                runInAction(() => {
                    this.employees = data;
                }),
            );
        } else {
            await getEmployeeList().then((data) =>
                runInAction(() => {
                    this.employees = data;
                }),
            );
        }

        runInAction(() => {
            this.loadingEmployees = false;
        });
    }

    closeModal(): void {
        this.isEditOpen = false;
    }

    openToEdit(employee: EmployeeEntity): void {
        //console.log(employee);
        const emp = this.employees.find((e) => e.id === employee.id);
        this.isEditOpen = true;
        if (emp) {
            runInAction(() => {
                this.employee = emp;
            });
        } else {
            throw new Error(`Zamestnanec s ID ${employee.id} nenalezen!`);
        }
    }

    async editEmployee(employee: EmployeeDto): Promise<void> {
        if (this.employee.id) {
            await updateEmployee(this.employee.id, employee) //TODO dat to ID pryc nejak
                .then(() => {
                    message.warning('Udaje o zamestnanci byly upraveny');
                })
                .catch((error) => {
                    console.log(error);
                    message.error('Nepodařilo se upravit zaměstnance');
                });
        }
        await this.fetchAllEmployees();
        runInAction(() => {
            this.closeModal();
        });
    }

    async addEmployee(employee: EmployeeDto): Promise<void> {
        await createEmployee(employee)
            .then(() => {
                message.success('Zaměstnanec byl úspěšně přidán');
            })
            .catch((error) => {
                console.log(error);
                message.error('Nepodařilo se vytvorit zaměstnance');
            });
        await this.fetchAllEmployees();
        runInAction(() => {
            this.closeModal();
        });
    }

    async deleteEmployee(id: number): Promise<void> {
        await deleteEmployee(id);
        runInAction(() => {
            this.loadingEmployees = true;
            message.success('Zamestnanec byl úspěšně smazán.');
        });
        await this.fetchAllEmployees();
        runInAction(() => {
            this.loadingEmployees = false;
        });
    }
}
