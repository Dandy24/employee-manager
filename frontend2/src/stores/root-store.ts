import { EmployeeStore } from './employee-store';
import { CompanyStore } from './company-store';
import { action, makeObservable, observable } from 'mobx';

export class RootStore {
    employeeStore: EmployeeStore;
    companyStore: CompanyStore;
    activePage = 'company-list';

    constructor() {
        this.employeeStore = new EmployeeStore(this);
        this.companyStore = new CompanyStore(this);

        makeObservable(this, {
            activePage: observable,

            setActivePage: action,
        });
    }

    setActivePage(newPage: string): void {
        this.activePage = newPage;
    }

    /*hydrate(): RootStoreHydration {
        return {};
    }

    rehydrate(data: RootStoreHydration): void {
        if (data) {
        }
    }*/
}
