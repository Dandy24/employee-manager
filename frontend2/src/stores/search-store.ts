import { RootStore } from './root-store';
import { action, makeObservable, observable } from 'mobx';
import { SearchResultItem } from '../components/search/search-result';
import { SearchableCompanyEntity } from '../models/entities/searchable-company-entity';
import { SearchableEmployeeEntity } from '../models/entities/searchable-employee-entity';

export class SearchStore {
    searchableCompanies: SearchableCompanyEntity[]; // TODO Entity SearchableItem
    searchableEmployees: SearchableEmployeeEntity[]; // TODO Entity SearchableItem
    loading = false;
    private rootStore: RootStore;

    constructor(rootStore: RootStore) {
        this.rootStore = rootStore;

        makeObservable(this, {
            searchableCompanies: observable,
            searchableEmployees: observable,
            loading: observable,

            createSearchableCompanies: action,
            setSearchableCompanies: action,
            createSearchableEmployees: action,
            setSearchableEmployees: action,
        });
    }

    setSearchableCompanies(companies: SearchableCompanyEntity[]): void {
        this.searchableCompanies = companies;
    }

    setSearchableEmployees(employees: SearchableEmployeeEntity[]): void {
        this.searchableEmployees = employees;
    }

    createSearchableCompanies = (): void => {
        const searchable = this.rootStore.companyStore.companies.map((company, index) => ({
            company,
            value: `${company.name} | ${company.phone} | ${company.address}`,
            label: SearchResultItem(company, index, this.rootStore),
            key: `company-${company.id}`,
        }));
        this.setSearchableCompanies(searchable);
    };

    createSearchableEmployees = (): void => {
        const searchable = this.rootStore.employeeStore.employees.map((employee, index) => ({
            employee,
            value: `${employee.first_name} ${employee.last_name} | ${employee.phone} | ${employee.company}`,
            label: SearchResultItem(employee, index, this.rootStore),
            key: `employee-${employee.id}`,
        }));
        this.setSearchableEmployees(searchable);
    };
}
