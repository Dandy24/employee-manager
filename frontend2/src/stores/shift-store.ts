import { RootStore } from './root-store';
import { action, makeObservable, observable, runInAction, toJS } from 'mobx';
import { EmployeeEntity } from '../models/entities/employee-entity';
import { createShift, getEmployeeListForShift, getShiftListForCompany, updateShift } from '../api/apiCalls';
import { ShiftEntity } from '../models/entities/shift-entity';
import { ShiftDto } from '../models/dtos/shift-dto';
import moment from 'moment';
import { ShiftTypeEnum } from '../models/enums/shift-type-enum';

export class ShiftStore {
    availableEmployees: EmployeeEntity[] = [];
    shiftEmployees: EmployeeEntity[];
    shift: ShiftEntity;
    shiftList: ShiftEntity[];

    isShiftSelectOpen = false;

    private rootStore: RootStore;

    constructor(rootStore: RootStore) {
        this.rootStore = rootStore;

        makeObservable(this, {
            isShiftSelectOpen: observable,
            setShiftSelectOpen: action,

            availableEmployees: observable,
            shiftEmployees: observable,
            shiftList: observable,

            loadShiftList: action,
            getShiftById: action,
            // getShiftForDateTime: action,
            addToShift: action,
            removeFromShift: action,
            clearShift: action,
            saveShift: action,
            setShiftEmployees: action,

            setEmployees: action,
            addEmployee: action,
            removeEmployee: action,
        });

        //this.availableEmployees = [...rootStore.employeeStore.employees];

        this.shiftEmployees = [];
    }

    setShiftSelectOpen(open: boolean): void {
        this.isShiftSelectOpen = open;
    }

    setShift(shiftId: number): void {
        this.shift = this.shiftList.find((shift) => shift.id === shiftId);
    }

    // getShiftForDateTime(date: string, time: any): ShiftEntity {
    //     return this.shiftList?.find((shift) => shift.date === date && shift.time == time);
    // }

    getShiftsForDate(date: string): ShiftEntity[] {
        return this.shiftList?.filter((shift) => shift.date === date);
    }

    getShiftById(shiftId: number): ShiftEntity {
        return this.shiftList.find((shift) => shift.id === shiftId);
    }

    async loadShiftList(companyId: number): Promise<void> {
        const shifts = await getShiftListForCompany(companyId);
        if (shifts) {
            runInAction(() => {
                this.shiftList = shifts;
            });
        }
    }

    addToShift(employee: EmployeeEntity, sourceIndex?: number, destinationIndex?: number): void {
        // if (this.shiftEmployees.length === 0) {
        this.shiftEmployees.push(employee);
        // } else {
        //     console.log(sourceIndex, destinationIndex);
        //     const items = [...this.shiftEmployees];
        //
        //     const [reorderedItem] = items.splice(sourceIndex, 1);
        //     items.splice(destinationIndex, 0, reorderedItem);
        //
        //     this.setShiftEmployees(items);
        // }
        //
        // console.log(this.shiftEmployees);
    }

    removeFromShift(index: number): void {
        this.shiftEmployees.splice(index, 1);
    }

    clearShift(): void {
        this.shiftEmployees = [];
    }

    setShiftEmployees(shift: EmployeeEntity[]): void {
        this.shiftEmployees = shift;
    }

    async saveShift(updatedShift: ShiftEntity): Promise<void> {
        // await this.rootStore.companyStore.fetchAllCompanies(); //TODO vyhodit to idealne a to radeji zkusit loadnout v useEffectu
        if (this.shift.id) {
            const employeeIDs = this.shiftEmployees.map((emp) => emp.id);
            const shift: ShiftDto = { ...updatedShift, employeeIDs };
            console.log(shift);
            // await updateShift(this.shift.id, updatedShift);
            await updateShift(this.shift.id, updatedShift);
        } else {
            const shift = new ShiftDto();
            shift.time = ShiftTypeEnum.Rano;
            shift.date = moment().format('YYYY-MM-DD');
            //shift.companyID = toJS(this.rootStore.companyStore.companies.find((comp) => comp.id === 3).id);
            shift.companyID = toJS(this.shift.companyID);
            // shift.employeeIDs = toJS(this.shiftEmployees.map((shift) => shift.id));
            shift.employeeIDs = this.shift.employeeIDs;
            await createShift(shift);
        }

        this.clearShift(); //TODO trigger this only if createShift was successfull (Exception wasnt thrown)
    }

    setEmployees(employees: EmployeeEntity[]): void {
        this.availableEmployees = employees;
    }

    addEmployee(employee: EmployeeEntity): void {
        this.availableEmployees.push(employee);
    }

    removeEmployee(index: number): void {
        this.availableEmployees.splice(index, 1);
    }

    async loadShiftEmployees(shiftId: number): Promise<void> {
        const loadedEmployees = await getEmployeeListForShift(shiftId);
        runInAction(() => {
            this.shiftEmployees = loadedEmployees;
        });
    }

    async loadAvailableEmployees(): Promise<void> {
        await this.rootStore.employeeStore.fetchAllEmployees(this.shift.companyID);
        this.setEmployees([...this.rootStore.employeeStore.employees]);
    }
}
