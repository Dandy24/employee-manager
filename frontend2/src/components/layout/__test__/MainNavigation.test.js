import { act, fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { MainNavigation } from '../MainNavigation';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import App from '../../../App';
import { RootStore } from '../../../stores/root-store';
import { RootStoreProvider } from '../../../stores/root-store-provider';

test('Menu snapshot matches the previous one', () => {
    const {asFragment} = render(
        <BrowserRouter>
            <MainNavigation />
        </BrowserRouter>,
    )

    expect(asFragment()).toMatchSnapshot();
});

test('Menu displays on page correctly', () => {
    const { getByTestId } = render(
        <BrowserRouter>
            <MainNavigation />
        </BrowserRouter>,
    );

    const mainMenuEl = getByTestId('main-header');

    expect(mainMenuEl).toBeInTheDocument();
});

test('Menu items display correctly in menu', () => {
    const { getByTestId } = render(
        <BrowserRouter>
            <MainNavigation />
        </BrowserRouter>,
    );

    const CompanyListEl = getByTestId('menu-company-list-item');

    expect(CompanyListEl.textContent).toBe('Seznam firem');
});

test('Menu items link correctly redirects to right page', () => {
    // Defining missing window.matchMedia() property from JSDOM, to prevent Jest from crashing
    // https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom

    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(), // deprecated
            removeListener: jest.fn(), // deprecated
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });

    const { getByTestId } = render(
        <RootStoreProvider rootStore={new RootStore()}>
            <BrowserRouter>
                <App />
            </BrowserRouter>
        </RootStoreProvider>,
    );

    const employeeLinkEl = getByTestId('menu-employee-list-item-link');

    act(() => {
        userEvent.click(employeeLinkEl);
    });

    expect(screen.getByText('ID zamÄ›stnance')).toBeInTheDocument();
});
