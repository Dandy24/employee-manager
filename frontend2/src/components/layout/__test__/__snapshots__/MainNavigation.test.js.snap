// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu snapshot matches the previous one 1`] = `
<DocumentFragment>
  <header
    class="ant-layout-header"
    data-testid="main-header"
  >
    <ul
      class="ant-menu-overflow ant-menu ant-menu-root ant-menu-horizontal ant-menu-dark"
      data-menu-list="true"
      role="menu"
      tabindex="0"
    >
      <li
        class="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child"
        data-menu-id="rc-menu-uuid-test-company-list"
        data-testid="menu-company-list-item"
        role="menuitem"
        style="opacity: 1; order: 0;"
        tabindex="-1"
      >
        <span
          class="ant-menu-title-content"
        >
          <a
            data-testid="menu-company-list-item-link"
            href="/"
          >
            Seznam firem
          </a>
        </span>
      </li>
      <li
        class="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child"
        role="menuitem"
        style="opacity: 1; order: 1;"
        tabindex="-1"
      >
        <span
          class="ant-menu-title-content"
        >
          <a
            href="/new-company"
          >
            Přidání nové firmy
          </a>
        </span>
      </li>
      <li
        class="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child"
        data-testid="menu-employee-list-item"
        role="menuitem"
        style="opacity: 1; order: 2;"
        tabindex="-1"
      >
        <span
          class="ant-menu-title-content"
        >
          <a
            data-testid="menu-employee-list-item-link"
            href="/employee-list"
          >
            Seznam zaměstnanců
          </a>
        </span>
      </li>
      <li
        class="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child"
        data-testid="menu-new-employee-item"
        role="menuitem"
        style="opacity: 1; order: 3;"
        tabindex="-1"
      >
        <span
          class="ant-menu-title-content"
        >
          <a
            data-testid="menu-new-employee-item-link"
            href="/new-employee"
          >
            Přidání nového zaměstnance
          </a>
        </span>
      </li>
      <li
        aria-hidden="true"
        class="ant-menu-overflow-item ant-menu-overflow-item-rest ant-menu-submenu ant-menu-submenu-horizontal"
        role="none"
        style="opacity: 0; height: 0px; overflow-y: hidden; order: 9007199254740991; pointer-events: none; position: absolute;"
      >
        <div
          aria-controls="rc-menu-uuid-test-rc-menu-more-popup"
          aria-expanded="false"
          aria-haspopup="true"
          class="ant-menu-submenu-title"
          data-menu-id="rc-menu-uuid-test-rc-menu-more"
          role="menuitem"
          tabindex="-1"
        >
          <span
            aria-label="ellipsis"
            class="anticon anticon-ellipsis"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="ellipsis"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
              />
            </svg>
          </span>
          <i
            class="ant-menu-submenu-arrow"
          />
        </div>
      </li>
    </ul>
    <div
      aria-hidden="true"
      style="display: none;"
    />
  </header>
</DocumentFragment>
`;
